### ISSUE TITLE
Network Errors, Connection Resets, Requiring Docker Restart

### ISSUE THREAD

**Issue by micchickenburger** [docker/for-mac#3448]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID:  70614E8F-4B41-4857-9044-8A1A2C1A5FFC/20190108175851

### Expected behavior

Networking between my Mac OS host and running docker containers should not sporadically stop working.

### Actual behavior

I'm experiencing network errors (connection resets, network errors) between my Mac host and all running docker containers.  Restarting the containers does not resolve the problem.  I have to restart docker to solve the problem.  I've experienced this problem twice today, but had never experienced it before.

### Information

  - macOS Version:  macOS Mojave 10.14.2

### Diagnostic logs

<img width="696" alt="image" src="https://user-images.githubusercontent.com/4521230/50850808-9929d100-1340-11e9-85c0-aad33b67f9b4.png">

### Steps to reproduce the behavior

It's hard to reproduce.  Connectivity just ceases seemingly randomly, requiring me to restart docker.  This is occurring on a fresh docker install as well.

**Comment by micchickenburger** [comment]
When the issue occurs, this is the error I receiving trying to run mongo on my host to connect with the monger server in my mongo docker container.

```
$ mongo
MongoDB shell version v4.0.2
connecting to: mongodb://127.0.0.1:27017
2019-01-08T12:18:13.494-0600 E QUERY    [js] Error: network error while attempting to run command 'isMaster' on host '127.0.0.1:27017'  :
connect@src/mongo/shell/mongo.js:257:13
@(connect):1:6
exception: connect failed
```

This is the issue I experienced trying to create an SQS queue on `localstack:localstack` running in a container:

```
Starting Localstack...
Started Localstack with hash 8a42a0e43174653894b4e1985d7d835f48b8da6ab46e7ecef57f66edaff4176d
Creating queue jobs at SQS endpoint http://localhost:4576/queue/jobs
An error occurred (502) when calling the CreateQueue operation (reached max retries: 4): Bad Gateway
```

Image versions:

```
$ docker image ls
REPOSITORY              TAG                 IMAGE ID            CREATED             SIZE
localstack/localstack   latest              c583aaf39486        4 days ago          1.02GB
mongo                   latest              7177e01e8c01        10 days ago         393MB
```

**Comment by lduchesne** [comment]
I too had this problem and downgrading to Docker Community Edition 18.06.0 seemed to have fixed it. It might be related to https://github.com/docker/for-mac/issues/3360

**Comment by schmurgon** [comment]
Same problem here on all versions higher than 18.06.1-ce-mac73 (26764).  For me, the network connections drop out when using the JVM debugger after about 60 seconds.

**Comment by ivester** [comment]
Same problem here

**Comment by jeff-cook** [comment]
I have noticed that when it happens the CPU spikes. I have not figured out if the spike is a cause or a symptom.

**Comment by josh-h** [comment]
@jeff-cook I'm not certain it is load related. I experienced similar symptoms with UDP packets failing to be bridged into the container. In each of the several cases the container ran for 24 hours or so and when returning to the office in the morning the networking to the container was dead. During the day the host would have constant use with periods of the CPU being driven hard. So, I suspect it is not load, but some other condition that triggers the bug.

My environment is a mac server with a static IP assigned, so it is not related to sleep or flaky wifi connections. Reverting to 18.06.1 seems to be a valid workaround, so far.

**Comment by jeff-cook** [comment]
It happened (76297123-260B-45B6-872E-9DE74FB5F950/20190111200619) even after a downgrade to Version 2.0.0.0-mac78 (28905) c404a62c3f

**Comment by josh-h** [comment]
This occurred again for me running `Server Version: 18.06.1-ce`.

I've determined that ingress network connections succeed (ie. I see UDP network packets arrive in the container). However, outbound connections from the container fail (ie. `ping google.com` doesn't receive any responses).

**Comment by Tobsucht** [comment]
Same here. As @josh-h already mentioned, I don't think it is load related.

macOS Mojave
Version 10.14

![bildschirmfoto 2019-01-23 um 12 59 20](https://user-images.githubusercontent.com/11056604/51605333-13e11780-1f0f-11e9-8462-758f84d2a6a0.png)

**Comment by chenxushuo** [comment]
Same problem here！！But ，after 30min i can get this error  when I restart docker service every time ！

**Comment by VillanCh** [comment]
me, too. I am using RabbitMQ, Postgres in Django App dev. After publishing or receiving some data, the network will be failed to my localhost, I cannot connect to rabbitmq and postgres from host. But I can still `docker-compose exec postgres bash` to enter the container to test, in the container, I can connect the db and mq.

Besides, I cannot ping the Postgres and RabbitMQ successfully.

After restart docker for mac, all recover. but a few moment later, it came to me again and again.....

I have to stop my coding to restart the docker, what a mess...

**Comment by MarounMaroun** [comment]
Same problem here. Does anyone have a solution already?

**Comment by elhay-av** [comment]
> Same problem here. Does anyone have a solution already?

https://github.com/docker/for-mac/issues/3448#issuecomment-452490002

**Comment by jezao** [comment]
Same problem here

**Comment by MarounMaroun** [comment]
@jezao Downgrading to 18.06.0-ce-mac70 2018-07-25 solved my problem. Download here: https://download.docker.com/mac/stable/26399/Docker.dmg

**Comment by daviyang35** [comment]
Same problem.

macOS 10.14.4
Docker Desktop 2.0.0.3(31259)

docker engine
```
Client: Docker Engine - Community
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        6247962
 Built:             Sun Feb 10 04:12:39 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.6
  Git commit:       6247962
  Built:            Sun Feb 10 04:13:06 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

docker-compose
```
docker-compose version 1.23.2, build 1110ad01
docker-py version: 3.6.0
CPython version: 3.6.6
OpenSSL version: OpenSSL 1.1.0h  27 Mar 2018
```

**Comment by MarounMaroun** [comment]
@daviyang35 Did you try to downgrade to **18.06.0-ce-mac70 2018-07-25**?

**Comment by daviyang35** [comment]
@MarounMaroun Yes. Use your link can ignore this issues. Thanks.

**Comment by keymandll** [comment]
I'm experiencing the same issue I guess. (Engine version 18.09.2)

Normally, it happens after a week of running like 10 containers. The load is always extremely light. (I have no numbers)
What I have noticed yesterday is that network traffic higher than usual (~200 transaction/second) results in losing the network connection to containers from the host. By one transaction I mean a set of connect, send, receive, disconnect operations.
Doing the exact same operations at 1 transaction/sec did not trigger the issue.

I remember seeing a log entry somewhere produced by docker about syn flood which could be related. Unfortunately I do not remember neither can figure out where I saw it.

I have noticed that when I experience the issue the networking between the containers still work fine.

**Comment by lucas-bremond** [comment]
Same issue. Seems to occur randomly. Restarting Docker fixes it.

**Comment by dmuth** [comment]
I had something similar (#3674) recently, and I ended up writing a test to reliably catch the issue, which I published at https://github.com/dmuth/docker-health-check

In my case, downgrading to 18.06.0-ce-mac70 also worked.

-- Doug

**Comment by Kaelten** [comment]
still an issue with the most recent release, downgrading seems to alleviate.

**Comment by anirudhwarrier** [comment]
Issue exists even on 2.0.0.3. Had to downgrade to 18.06.0-ce-mac70.

**Comment by wilomgfx** [comment]
Same here, i keep getting network issues. Containers are running fine, i can connect to them via bash, but they won't load on my browser or connect to each other.

**Comment by rubnov** [comment]
I've been banging my head against the wall with this issue. I'm running a Django app in docker container, which needs to connect to a Postgres database on the host machine. I am getting the following error quite often (every 2 - 10 minutes):

```
django.db.utils.OperationalError: could not connect to server: Connection timed out
```


I've downgraded from Docker 2.1.0.3 (currently the latest) to 18.06.0-ce-mac70 as suggested above, and the error above disappeared, but only to be replaced by this error, which is now occurring even more often: 

```
django.db.utils.OperationalError: could not connect to server: Connection refused
```

This seems to be related to load / number of requests. The issue occurs more often in high load situations. 

Any suggestions on what I can do next?

**Comment by kiragaz** [comment]
Problem is still reproducible on 2.1.0.5

**Comment by dko-slapdash** [comment]
@guillaumerose could you please help with this issue? Basically, docker for mac can't be reliably used for development in some cases, the containers just stop responding on network connections.

- A lot of people complain for >1Y.
- People even start writing TOOLS to work-around this issue by restarting Docker (with AppleScript).
- @dmuth has clear steps to reproduce it in https://github.com/docker/for-mac/issues/3674 and https://github.com/dmuth/docker-health-check
- All engineers in our company suffer from something which looks very similar to the current issue: we have to restart a container with elasticsearch time few times a day.

E.g. right now I have the following symptoms for elasticsearch container listening on port 9200:

```
$ curl -vvv http://127.0.0.1:9200/zzz
<fails when running from Mac>
curl: (7) Failed to connect to 127.0.0.1 port 9200: Operation timed out

$ netstat -na | grep 9200
<shows 128 CLOSE_WAIT connections which are stale forever>
tcp4     271      0  127.0.0.1.9200         127.0.0.1.59449        CLOSE_WAIT
tcp4     271      0  127.0.0.1.9200         127.0.0.1.59446        CLOSE_WAIT
tcp4       0      0  127.0.0.1.9200         *.*                    LISTEN

$ pgrep node
<shows nothing, i.e. the connecting processes died long time ago>

$ sudo tcpdump -i any -n -p tcp port 9200 & curl -vvv http://127.0.0.1:9200/zzz
<only SYN packets travel?>
02:09:59.303560 IP 127.0.0.1.63681 > 127.0.0.1.9200: Flags [S], seq 1329676029, win 65535, options [mss 16344,nop,wscale 6,nop,nop,TS val 664304244 ecr 0,sackOK,eol], length 0
...
02:10:02.815838 IP 127.0.0.1.63681 > 127.0.0.1.9200: Flags [S], seq 1329676029, win 65535, options [mss 16344,nop,wscale 6,nop,nop,TS val 664307746 ecr 0,sackOK,eol], length 0
02:10:06.022543 IP 127.0.0.1.63681 > 127.0.0.1.9200: Flags [S], seq 1329676029, win 65535, options [mss 16344,sackOK,eol], length 0
...
02:10:18.867903 IP 127.0.0.1.63681 > 127.0.0.1.9200: Flags [S], seq 1329676029, win 65535, options [mss 16344,sackOK,eol], length 0

$ docker-compose exec elasticsearch bash
# curl http://127.0.0.1:9200/
<succeeds from inside the container>
{
  "name" : "41d967fa27a4",
  "cluster_name" : "docker-cluster",
```

**Comment by piniondna** [comment]
Has this issue been officially addressed or acknowledged at all? I’ve personally had multiple colleagues experience this same issue, and it seems to appear more with relatively complicated/large stacks.

It’s also complicating my efforts to convert developers into Docker uses when they experience issues like this that shake their confidence in the technology.

The fact that there hasn’t been any official acknowledgement of this problem strange. I think any number of devs (myself included) would be more than happy to help diagnose the root cause if called upon, but that hasn’t happened in more than a year?

Is Docker still being developed as a product? Or has is secretly gone into maintenance mode?

**Comment by slikts** [comment]
Note that [downgrading](https://github.com/docker/for-mac/issues/3448#issuecomment-478936638) should probably still be a viable workaround.

**Comment by thtas** [comment]
Just anecdotal... But I suffered with this issue for a while, trying a new version every few months and then downgrading back to version i knew worked. Eventually somebody suggested it might be a memory issue, so i upgraded again and gave Docker a heap more memory (currently allocated 12gb) and the problem went away. It's been solid for a few months now.

**Comment by piniondna** [comment]
Adding resources helped for me up to a point, but wasn’t a panacea.

Even if the “fix” is just surfacing an error message to the user, this is better than a silent failure leading to hours of app troubleshooting, just to find out it’s a failure with docker.

**Comment by coding-bunny** [comment]
This issue is still a big pain.
My ElasticSearch container randomly spits out the error about a GET call failed and the entire docker crashes and needs to be restarted.
This makes development really difficult.

**Comment by dko-slapdash** [comment]
BTW I mitigated it a little by just opening much less connections to ES and making those connections persistent.

It looks like there is some memory (or connections?) leak in docker port forwarding proxy, so if you open and close a lot of concurrent connections from host to a container and do it many times per second, eventually these connections leak (**even when the process which opens connections is terminated** - so it's really a bug in docker-desktop, not in the calling app). I posted netstat/tcpdump above.

Before, we were accidentally opening one new connection per ES query and then closing them at random moments of time (sometimes keeping them open for a long time). After we switched to persistent connections, the docker bug disappeared too.

**Comment by flow3d** [comment]
This issue is open for over a year without any viable workaround (installing a 2018 version doesn't work, at least for me).

Any other suggestions?

**Comment by Talhazzers** [comment]
Same problem.

**Comment by rndnoise** [comment]
I have the same problem. At least I'm not alone, I guess.

**Comment by danielkihlgren** [comment]
I'm having this issue too. Are there any official statements regarding this issue?
I'm using v 3.1.0 @ MacOs Catalina 10.15.7
I can't ping anything on my network from within the container. I can reach other services within the container. I'm also able to connect to the container from the outside using the exposed ports.

**Comment by simonjthomas** [comment]
Confirming the same issue. It's only started relatively recently, I think after I added a new container to a compose file but I'm not certain as the timing doesn't line up perfectly. I've just thrown a load more resources at docker to see if it helps, but it had plenty already. Docker 3.1.0 on MacOS 10.15.7.

I can access the container's web server through a HTTP connection, but the containers can't communicate with each other and can't talk to the outside world (pinging an external site from within the container doesn't work). It works fine for about two days at a time before entering this state with no config changes. restarting containers doesn't help, only restarting docker or the machine docker's running on.

**Comment by christianhuening** [comment]
same here, Docker 3.3.1 on Big Sur 11.2.3

**Comment by npoczynek** [comment]
Seeing the same issue with Docker 3.3.3 on Big Sur 11.3.1. I suspect @dko-slapdash is on the right track re: resource leak related to number or rate of connections; running an aggressive nmap scan from a container to the host results in almost immediate network failures that are only resolved by restarting Docker engine.

**Comment by marcellp** [comment]
Pinging some maintainers here: @StefanScherer @djs55 @stephen-turner in case this issue flew under their radars.

As a summary of this thread, users have been experiencing intermittent network failures, connection resets, etc., requiring a full restart of Docker or pruning all images and networks for the thing to come back on-line. Seems to be related to a resource/memory leak, and happens more often under heavy load spikes and adding more resources to Docker seems to work for a little bit. There seem to be relatively easy steps to reproduce this problem at will, see e.g. https://github.com/docker/for-mac/issues/3448#issuecomment-628507263. 

We have been experiencing similar problems with Docker for Mac for the past 6 months and our team is often losing hours of productivity because of this issue. Would really appreciate it if one of you could look into this.

**Comment by rs-solutions** [comment]
Same issue with version 3.5.2 on Big Sur

**Comment by flow3d** [comment]
If the problem is reproducing, I'll be happy to go on a call with any1 willing to tackle this

**Comment by garylimkc** [comment]
yes, having the same issue here. Using Docker 3.5.2 and previously network was working within container but suddenly unable to connect.

**Comment by bradleyayers** [comment]
I spent some time looking into this and narrowed it down to a simple reproduction, I actually posted these on a docker-for-windows issue https://github.com/docker/for-win/issues/8861#issuecomment-903421446. Basically running a python HTTP server and using siege as the HTTP client can reproduce the issue in ~7 minutes and ~15k requests.

**Comment by pich4ya** [comment]
I can confirm that Docker Engine 20.10.8 is still having this issue. Containers cannot reach host's network if one of the them sent lots network packets.

**Comment by bradleyayers** [comment]
I ended up working around this issue by using long live connections between my container and host, rather than many short live connections. Specifically I was using an nginx container as a reverse proxy and had the `Connection: close` header set which essentially prevented socket reuse.

The issue seems to be with the number of sockets that are established, rather than the amount of data transferred.

**Comment by pich4ya** [comment]
I finally got the fcking working solution:

1. Turn on "Experimental Feature" > "Use the new Virtualization framework"
2. Edit "/Users/$USER/Library/Group Containers/group.com.docker/settings.json"
```
  "useVpnkit": false,
```
3. Restart Docker Daemon

You can thank me later.

**Comment by flow3d** [comment]
@pich4ya didn't solve my problem :/

**Comment by flow3d** [comment]
TL;DR - connect to the docker using your public facing IP rather than 127.0.0.1

Had some progress :)
realized that most of the time docker prefers to use the kernel's routing / NAT service to do its networking. However, for loopback traffic that can't be routed, they use docker-proxy. Hence, since the connection is established and the data isn't received, I suspected the user-level component to mess things up.

Changed my connection from localhost to the device's public IP - no more issues. Back to 127.0.0.1 - issues are back.

To any maintainer who may be tracking this issue - docker-proxy seems to be the one causing the issues :(

**Comment by stevenleeg** [comment]
> TL;DR - connect to the docker using your public facing IP rather than 127.0.0.1
> 
> Had some progress :) realized that most of the time docker prefers to use the kernel's routing / NAT service to do its networking. However, for loopback traffic that can't be routed, they use docker-proxy. Hence, since the connection is established and the data isn't received, I suspected the user-level component to mess things up.
> 
> Changed my connection from localhost to the device's public IP - no more issues. Back to 127.0.0.1 - issues are back.
> 
> To any maintainer who may be tracking this issue - docker-proxy seems to be the one causing the issues :(

Just want to confirm that I'm also running into this issue on an M1 Macbook Air and found that connecting via my computer's IP address instead of `localhost` fixed the problem.

**Comment by merges** [comment]
Same problem. Is this being looked at by the maintainers at all?

**Comment by razorman8669** [comment]
seems like the for-win folks have a very similar issue as the for-mac peeps.  I used a suggestion from this other thread https://github.com/docker/for-win/issues/8861 and I can confirm that by downgrading docker desktop for mac to version 4.5.0 "fixes" the problem and I no longer have network issues. I tested for several days without any problems. Then upgraded back to the latest version and started getting timeouts and connection failures almost immediately again.

there must have been something that occurred between 4.5.0 and now that is causing networking issues.

**Comment by alisterscott** [comment]
> I finally got the fcking working solution:

>   Turn on "Experimental Feature" > "Use the new Virtualization framework"
>  Edit "/Users/$USER/Library/Group Containers/group.com.docker/settings.json"
>  "useVpnkit": false,
>  Restart Docker Daemon

> You can thank me later.

@pich4ya works on Macbook Air M1 - thank you 👍

**Comment by fabswt** [comment]
The option mentioned in the previous post is now found in the Docker preferences in *General* under:

> Use Virtualization framework

Seems to help, it's been 24 hours since I ticked the option and no more issues, while I'd typically lose connectivity 3 to 5 times a day.

My issue was: I'd regularly lose the connection to some APIs from inside Docker. Regularly, my Python Flask app would no longer be able to connect to the OpenAI API (timeout) from Docker, curl from Docker would not be able either (timeout); BUT: curl calls from macOS would still work (so not an issue with the API), as would API calls to Azure from inside the Docker container. Restarting the container did not help. Had to restart Docker every time to restore proper connectivity. Did not know how to diagnose the issue.

**Comment by fabswt** [comment]
So, doing some local development… In 3 hours of coding, had to restart the Docker app 10 times, because it'd lost connectivity as described in the previous post. 

I love dev-prod parity but it's getting counter-productive.

**Comment by fabswt** [comment]
With useVpnkit=false takes forever to start. Tested on 4.15.0 to 4.16.1.

**Comment by maxgurewitz** [comment]
On macOS Monterey 12.6.2, docker version 20.10.22, and I'm experiencing this issue. Tried setting `"useVpnkit": false` but docker still hadn't re-started after 15 minutes, so I gave up on that approach.

**Comment by iancmcc** [comment]
I'd been disabling vpnkit to get around this issue for months, but as of Ventura 13.2 / Docker Desktop 4.16.2, docker will no longer start with `"useVpnkit": false` and this issue is as present as ever. Not sure what to do at this point; Docker is effectively useless for any stack that opens a significant number of connections on startup.

**Comment by iancmcc** [comment]
Downgrading to Docker Desktop 4.8.2 and reapplying the `"useVpnkit": false` configuration works around the issue for now.

**Comment by dimitrovs** [comment]
I was able to solve this problem by removing Docker Desktop for Mac and installing:
`Bash
brew install docker docker-compose colima
colima start --cpu 1 --memory 2 --disk 10
`
I also had to edit `~/.docker/config.json` to remove the credsStore line.
After that everything is working correctly.

**Comment by robevansuk** [comment]
Switching to Orbstack resolved this issue for me. However I wanted to add a few things I tried which may or may not be helpful...

Running `docker compose up` with a single postgres db service - worked as expected I could connect to localhost:5432 without issue.
Running `docker compose up` with 2 services (or more) - network connectivity issues experienced. 10-20% success rate using the same localhost:5432 connection url as before.

I deleted containers, volumes, networks between runs using: 

```
docker rm -f $(docker ps -aq); docker volume rm $(docker volume ls -qf dangling=true); docker network prune -f;
```

orbstack didn't seem to have the same issue mapping the container across the host network and was 100% reliable in comparison to docker-desktop which is essentially now obsolete for me as a result of its failure to reliably map containers to the relevant host/port so I can connect from the IDE to the database. Other containers didn't seem to have an issue as I was also running PgWeb as well as Postgres

**Comment by pkit** [comment]
Just another 5 cents.
If using Docker Desktop: pinning it to 1 CPU thread removes all problems (obviously at the expense of being slow asf), so it's a clear multithreading bug.
When using Colima it mostly works. Why mostly: still get random "connection reset" from time to time, pretty rarely though.
I suspect it's just OSX problem after all though because why would it be that similar for docker desktop and colima?
Obviously on my Linux dev machines nothing ever flaked in these tests.
